File: lib/services/github_csv_service.dart
import 'dart:convert';
import 'dart:io';
import 'package:http/http.dart' as http;
import 'package:flutter/services.dart';
import 'package:flutter/foundation.dart';

class GitHubCSVService {
  // GitHub Raw API base URL - kullanıcının repo'su
  static const String _baseUrl =
      'https://raw.githubusercontent.com/kaboya19/WebTufeMobile/main';

  // CORS proxy for web (Flutter web can't directly access GitHub due to CORS)
  static const String _corsProxy = 'https://api.allorigins.win/raw?url=';

  // Cache için timeout süresi (dakika)
  static const int _cacheTimeoutMinutes = 15;

  // Cache verilerini tutmak için static map
  static final Map<String, _CachedData> _cache = {};

  // Dosya adı alternatifleri - ASCII ve Turkish karakterli versiyonlar
  static final Map<String, List<String>> _fileNameAlternatives = {
    'ozel_gostergeler.csv': ['ozel_gostergeler.csv', 'özelgöstergeler.csv'],
    'ozelgostergeleraylik.csv': [
      'ozelgostergeleraylik.csv',
      'özelgöstergeleraylık.csv'
    ],
    'maddeleraylik.csv': ['maddeleraylik.csv', 'maddeleraylık.csv'],
    'harcama_gruplariaylik.csv': [
      'harcama_gruplariaylik.csv',
      'harcama_gruplarıaylık.csv'
    ],
    'harcama_gruplari.csv': ['harcama_gruplari.csv', 'harcama_grupları.csv'],
    'urunler.csv': ['urunler.csv', 'ürünler.csv'],
    'tufe.csv': ['tufe.csv', 'tüfe.csv'],
  };

  /// Cache busting için tarih parametresi oluşturur (format: ?v=YYYYMMDDHHMM)
  static String _getCacheBustingParam() {
    final now = DateTime.now();
    return '?v=${now.year}${now.month.toString().padLeft(2, '0')}${now.day.toString().padLeft(2, '0')}${now.hour.toString().padLeft(2, '0')}${now.minute.toString().padLeft(2, '0')}';
  }

  /// GitHub'dan CSV dosyasını çeker
  /// [fileName] - assets klasöründeki dosya adı (örn: 'gruplaraylik.csv')
  /// [useCache] - Cache kullanılsın mı (varsayılan: true)
  static Future<String> loadCSVFromGitHub(String fileName,
      {bool useCache = true}) async {
    // Cache kontrolü
    if (useCache && _cache.containsKey(fileName)) {
      final cachedData = _cache[fileName]!;
      if (DateTime.now().difference(cachedData.timestamp).inMinutes <
          _cacheTimeoutMinutes) {
        print('Cache\'den CSV okunuyor: $fileName');
        return cachedData.data;
      } else {
        // Cache süresi dolmuş, kaldır
        _cache.remove(fileName);
      }
    }

    // Dosya adı alternatifleri ile dene
    List<String> filesToTry = _fileNameAlternatives[fileName] ?? [fileName];

    for (String tryFileName in filesToTry) {
      try {
        print('Denenen dosya: $tryFileName');

        // Platform'a göre URL oluştur
        String url;
        if (kIsWeb) {
          // Web için jsDelivr CDN kullan (en güvenilir)
          url =
              'https://cdn.jsdelivr.net/gh/kaboya19/WebTufeMobile@main/assets/$tryFileName${_getCacheBustingParam()}';
          print('jsDelivr CDN ile CSV çekiliyor (ana method): $url');
          print('Dosya adı: [$tryFileName]');
        } else {
          // Mobil için direkt GitHub URL
          url = '$_baseUrl/assets/$tryFileName';
          print('GitHub\'dan CSV çekiliyor: $url');
        }

        final response = await http.get(
          Uri.parse(url),
          headers: {
            'Accept': 'text/csv',
            'Cache-Control': 'no-cache',
          },
        ).timeout(const Duration(seconds: 30));

        if (response.statusCode == 200) {
          // UTF-8 decode et
          String csvData;
          if (response.headers['content-type']?.contains('charset=utf-8') ==
              true) {
            csvData = response.body;
          } else {
            csvData = utf8.decode(response.bodyBytes);
          }

          print(
              'CSV başarıyla yüklendi: $tryFileName (${csvData.length} karakter)');

          // Debug: İlk birkaç satırı logla
          final lines = csvData.split('\n');
          print('CSV ilk satır: ${lines.isNotEmpty ? lines[0] : 'BOŞ'}');
          print('CSV satır sayısı: ${lines.length}');
          if (lines.length > 1) {
            print('CSV ikinci satır: ${lines[1]}');
          }

          // Cache'e kaydet
          if (useCache) {
            _cache[fileName] = _CachedData(csvData, DateTime.now());
          }

          return csvData;
        } else {
          print('HTTP hatası $tryFileName için: ${response.statusCode}');
          // Bu dosya adı başarısız, bir sonrakini dene
          continue;
        }
      } on SocketException catch (e) {
        print('İnternet bağlantısı hatası $tryFileName için: $e');
        continue;
      } on http.ClientException catch (e) {
        print('HTTP istek hatası $tryFileName için: $e');
        continue;
      } catch (e) {
        print('GitHub CSV yükleme hatası $tryFileName için: $e');
        continue;
      }
    }

    // Hiçbir dosya adı çalışmazsa fallback dene
    print('Tüm dosya adları başarısız, fallback deneniyor: $fileName');
    return await _loadLocalFallback(fileName);
  }

  /// Fallback mekanizması - alternatif yöntemler dener
  static Future<String> _loadLocalFallback(String fileName) async {
    if (kIsWeb) {
      // Web için alternatif CORS proxy dene
      try {
        print('Alternatif CORS proxy deneniyor: $fileName');
        final githubUrl = '$_baseUrl/assets/$fileName';
        final altProxyUrl =
            'https://corsproxy.io/?${Uri.encodeComponent(githubUrl)}';

        final response = await http.get(
          Uri.parse(altProxyUrl),
          headers: {'Accept': 'text/csv'},
        ).timeout(const Duration(seconds: 30));

        if (response.statusCode == 200) {
          print('Alternatif proxy başarılı: $fileName');
          return utf8.decode(response.bodyBytes);
        }
      } catch (e) {
        print('Alternatif proxy da başarısız: $e');
      }

      // Son çare: jsDelivr CDN dene
      try {
        print('jsDelivr CDN deneniyor: $fileName');
        final cdnUrl =
            'https://cdn.jsdelivr.net/gh/kaboya19/WebTufeMobile@main/assets/$fileName${_getCacheBustingParam()}';

        final response = await http.get(
          Uri.parse(cdnUrl),
          headers: {'Accept': 'text/csv'},
        ).timeout(const Duration(seconds: 30));

        if (response.statusCode == 200) {
          print('jsDelivr CDN başarılı: $fileName');
          final csvData = utf8.decode(response.bodyBytes);

          // Debug: İlk birkaç satırı logla
          final lines = csvData.split('\n');
          print(
              'CSV ilk satır (fallback): ${lines.isNotEmpty ? lines[0] : 'BOŞ'}');
          print('CSV satır sayısı (fallback): ${lines.length}');
          if (lines.length > 1) {
            print('CSV ikinci satır (fallback): ${lines[1]}');
          }

          return csvData;
        }
      } catch (e) {
        print('jsDelivr CDN da başarısız: $e');
      }

      // Son çare olarak Turkish karakterli dosyaları dene
      List<String> filesToTry = _fileNameAlternatives[fileName] ?? [fileName];

      for (String tryFileName in filesToTry) {
        if (tryFileName == fileName) continue; // Zaten denedik

        try {
          print('Fallback Turkish karakterli dosya deneniyor: $tryFileName');
          final cdnUrl =
              'https://cdn.jsdelivr.net/gh/kaboya19/WebTufeMobile@main/assets/$tryFileName${_getCacheBustingParam()}';

          final response = await http.get(
            Uri.parse(cdnUrl),
            headers: {'Accept': 'text/csv'},
          ).timeout(const Duration(seconds: 30));

          if (response.statusCode == 200) {
            print('Turkish karakterli dosya başarılı: $tryFileName');
            return utf8.decode(response.bodyBytes);
          }
        } catch (e) {
          print('Turkish karakterli dosya başarısız: $tryFileName - $e');
        }
      }
    }

    // Yerel asset'i dene (son çare)
    try {
      print('Yerel asset deneniyor: $fileName');
      return await rootBundle.loadString('assets/$fileName');
    } catch (e) {
      print('Yerel asset da başarısız: $e');
      throw Exception('$fileName dosyası hiçbir yöntemle yüklenemedi');
    }
  }

  /// Cache verilerini temizle
  static void clearCache() {
    _cache.clear();
    print('CSV cache temizlendi');
  }

  /// Belirli bir dosyanın cache'ini temizle
  static void clearFileCache(String fileName) {
    _cache.remove(fileName);
    print('$fileName cache\'i temizlendi');
  }

  /// İnternet bağlantısı kontrolü
  static Future<bool> hasInternetConnection() async {
    try {
      final result = await InternetAddress.lookup('google.com');
      return result.isNotEmpty && result[0].rawAddress.isNotEmpty;
    } on SocketException catch (_) {
      return false;
    }
  }
}

/// Cache verisi için yardımcı sınıf
class _CachedData {
  final String data;
  final DateTime timestamp;

  _CachedData(this.data, this.timestamp);
}

File: src/services/GitHubCSVService.ts
import AsyncStorage from '@react-native-async-storage/async-storage';
import {Platform} from 'react-native';

const CACHE_TIMEOUT_MINUTES = 60; // 1 saat cache
const CACHE_PREFIX = 'csv_cache_';
const CACHE_TIMESTAMP_PREFIX = 'csv_timestamp_';

// Yerleşik katkı CSV fallback (remote/asset bulunamazsa)
const LOCAL_KATKI_CSV = `,Alkollü içecekler ve tütün,Ev eşyası,Eğitim,Eğlence ve kültür,Giyim ve ayakkabı,Gıda ve alkolsüz içecekler,Haberleşme,"Konut,Su,Elektrik,Gaz ve Diğer Yakıtlar",Lokanta ve oteller,Sağlık,Ulaştırma,Çeşitli mal ve hizmetler,Web TÜFE
2025-02-28,0.0,0.48,0.17,0.07,0.35,0.96,-0.04,0.82,0.31,-0.31,0.27,0.2,3.29
2025-03-31,0.38,0.29,0.18,0.03,0.52,1.44,-0.05,0.39,0.09,-0.01,0.02,0.19,3.45
2025-04-30,0.0,0.07,0.0,0.02,0.09,0.51,-0.07,0.83,0.36,0.0,0.42,0.11,2.34
2025-05-31,0.0,0.13,0.0,0.03,-0.0,-0.11,-0.0,0.59,0.4,0.0,0.3,0.05,1.39
2025-06-30,0.0,0.04,0.0,0.01,0.04,-0.06,-0.01,0.52,0.33,-0.0,0.37,0.06,1.3
2025-07-31,0.2,0.11,0.1,0.02,0.05,0.19,0.01,0.94,0.22,0.0,0.44,0.04,2.32
2025-08-31,0.26,0.11,-0.0,0.02,-0.29,0.58,0.06,0.39,0.37,-0.0,0.4,0.06,1.94
2025-09-30,0.01,0.07,0.66,0.02,0.1,0.78,0.07,0.48,0.14,0.0,0.17,0.12,2.61
2025-10-31,0.07,0.1,0.01,0.02,0.89,1.14,0.0,0.39,0.08,0.0,0.25,0.13,3.09
2025-11-30,0.14,-0.0,-0.0,-0.0,-0.08,0.41,0.04,0.34,0.05,-0.0,0.24,0.0,1.14
2025-12-31,-0.0,0.1,0.0,0.04,0.04,0.43,0.1,0.27,0.09,0.04,-0.16,0.06,1.01
`;

const fileNameAlternatives: {[key: string]: string[]} = {
  'ozel_gostergeler.csv': ['ozel_gostergeler.csv', 'özelgöstergeler.csv'],
  'ozelgostergeleraylik.csv': [
    'ozelgostergeleraylik.csv',
    'özelgöstergeleraylık.csv',
  ],
  'ozelgostergeleryıllık.csv': [
    'ozelgostergeleryıllık.csv',
    'özelgöstergeleryıllık.csv',
    'ozelgostergeleryillik.csv',
  ],
  'maddeleraylik.csv': ['maddeleraylik.csv', 'maddeleraylık.csv'],
  'maddeleryıllık.csv': [
    'maddeleryıllık.csv',
    'maddeleryıllık.csv',
    'maddeleryillik.csv',
  ],
  'harcama_gruplariaylik.csv': [
    'harcama_gruplariaylik.csv',
    'harcama_gruplarıaylık.csv',
  ],
  'harcamagruplarıyıllık.csv': [
    'harcamagruplarıyıllık.csv',
    'harcamagruplariyillik.csv',
  ],
  'harcama_gruplari.csv': [
    'harcama_gruplari.csv',
    'harcama_grupları.csv',
    'harcamagrupları.csv',
  ],
  'gruplaryıllık.csv': [
    'gruplaryıllık.csv',
    'gruplaryillik.csv',
  ],
  'urunler.csv': ['urunler.csv', 'ürünler.csv'],
  'tufe.csv': ['tufe.csv', 'tüfe.csv'],
  'tüfeyıllık.csv': [
    'tüfeyıllık.csv',
    'tufeyillik.csv',
    'tufeyillik.csv',
    'tufe_yillik.csv',
  ],
  'katkıpayları.csv': ['katkıpayları.csv', 'katkipaylari.csv', 'katki_paylari.csv'],
};

export class GitHubCSVService {
  static async loadCSVFromGitHub(
    fileName: string,
    useCache: boolean = true
  ): Promise<string> {
    // Cache kontrolü
    if (useCache) {
      const cachedData = await this.getCachedData(fileName);
      if (cachedData) {
        // Cache'den okundu (log azaltıldı)
        return cachedData;
      }
    }

    // Dosya adı alternatifleri ile dene
    const filesToTry = fileNameAlternatives[fileName] || [fileName];

    for (const tryFileName of filesToTry) {
      try {
        // React Native'de assets dosyalarını okumak için
        // require kullanarak veya fetch ile okuyabiliriz
        // Burada örnek olarak fetch kullanıyoruz, gerçek implementasyonda
        // assets klasöründen okunması gerekiyor
        
        // Not: React Native'de assets dosyalarını okumak için
        // genellikle require() kullanılır veya bundle içinde olmalıdır
        // Bu örnekte fetch ile GitHub'dan okuma simüle ediyoruz
        
        // Gerçek implementasyonda:
        // const csvData = require(`../assets/${tryFileName}`);
        // veya
        // const csvData = await fetch(`file:///android_asset/${tryFileName}`);
        
        // Şimdilik placeholder olarak boş string döndürüyoruz
        // Gerçek implementasyonda bu kısım düzeltilmeli
        const csvData = await this.loadAssetFile(tryFileName);

        console.log(`CSV yüklendi: ${tryFileName}`);

        // Cache'e kaydet
        if (useCache) {
          await this.setCachedData(fileName, csvData);
        }

        return csvData;
      } catch (e) {
        // Yerel asset okuma hatası (log azaltıldı)
        continue;
      }
    }

    throw new Error(`${fileName} dosyası yerel assets'te bulunamadı`);
  }

  private static async loadAssetFile(fileName: string): Promise<string> {
    // GitHub'dan CSV dosyalarını okuma
    // Flutter versiyonunda assets klasöründen okunuyordu,
    // React Native versiyonunda GitHub'dan fetch ile okuyoruz
    const baseUrl = 'https://raw.githubusercontent.com/kaboya19/WebTufeMobile/main/assets/';
    
    // Önce ana dosya adını dene
    const filesToTry = fileNameAlternatives[fileName] || [fileName];
    
    for (const tryFileName of filesToTry) {
      try {
        const response = await fetch(`${baseUrl}${encodeURIComponent(tryFileName)}`);
        
        if (response.ok) {
          const text = await response.text();
          console.log(`CSV GitHub'dan yüklendi: ${tryFileName}`);
          return text;
        }
      } catch (e) {
        // Bu dosya adı çalışmadı, bir sonrakini dene
        continue;
      }
    }
    
    // Katkı CSV için fallback (sadece GitHub'dan okunamazsa)
    if (fileNameAlternatives['katkıpayları.csv']?.includes(fileName)) {
      console.log(`Katkı CSV GitHub'dan okunamadı, fallback kullanılıyor`);
      return LOCAL_KATKI_CSV;
    }
    
    // Hiçbiri çalışmadıysa hata fırlat
    console.log(`GitHub'dan CSV okuma hatası ${fileName} için: Dosya bulunamadı (404)`);
    throw new Error(`CSV dosyası GitHub'da bulunamadı: ${fileName}. Lütfen dosyayı GitHub repository'nize yükleyin.`);
  }

  private static async getCachedData(fileName: string): Promise<string | null> {
    try {
      const cacheKey = `${CACHE_PREFIX}${fileName}`;
      const timestampKey = `${CACHE_TIMESTAMP_PREFIX}${fileName}`;
      
      const cachedData = await AsyncStorage.getItem(cacheKey);
      const timestampStr = await AsyncStorage.getItem(timestampKey);
      
      if (!cachedData || !timestampStr) {
        return null;
      }

      const timestamp = parseInt(timestampStr, 10);
      const now = Date.now();
      const diffMinutes = (now - timestamp) / (1000 * 60);

      if (diffMinutes < CACHE_TIMEOUT_MINUTES) {
        return cachedData;
      } else {
        // Cache süresi dolmuş, kaldır
        await AsyncStorage.multiRemove([cacheKey, timestampKey]);
        return null;
      }
    } catch (e) {
      // Cache okuma hatası (sessiz)
      return null;
    }
  }

  private static async setCachedData(
    fileName: string,
    data: string
  ): Promise<void> {
    try {
      const cacheKey = `${CACHE_PREFIX}${fileName}`;
      const timestampKey = `${CACHE_TIMESTAMP_PREFIX}${fileName}`;
      
      await AsyncStorage.multiSet([
        [cacheKey, data],
        [timestampKey, Date.now().toString()],
      ]);
    } catch (e) {
      // Cache yazma hatası (sessiz)
    }
  }

  static async clearCache(): Promise<void> {
    try {
      const keys = await AsyncStorage.getAllKeys();
      const cacheKeys = keys.filter(
        key =>
          key.startsWith(CACHE_PREFIX) || key.startsWith(CACHE_TIMESTAMP_PREFIX)
      );
      await AsyncStorage.multiRemove(cacheKeys);
      console.log('CSV cache temizlendi');
    } catch (e) {
      console.log('Cache temizleme hatası:', e);
    }
  }

  static async clearFileCache(fileName: string): Promise<void> {
    try {
      const cacheKey = `${CACHE_PREFIX}${fileName}`;
      const timestampKey = `${CACHE_TIMESTAMP_PREFIX}${fileName}`;
      await AsyncStorage.multiRemove([cacheKey, timestampKey]);
      console.log(`${fileName} cache'i temizlendi`);
    } catch (e) {
      console.log('Dosya cache temizleme hatası:', e);
    }
  }
}

